File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
import sys

N = int(sys.stdin.readline())
score = []
for _ in range(N):
    score.append(list(map(int, sys.stdin.readline().split())))

# True면 스타트팀 선수
visited = [False]*N

answer = sys.maxsize

# n단계
def backtracking(n, idx):
    global answer
    # 팀 구성이 완료되었으면
    if n==N//2:
        #점수 추가하기
        start = 0
        rink = 0
        for i in range(N):
            # 중첩 방지
            for j in range(i+1, N):
                # 둘 다 스타트팀
                if visited[i] and visited[j]:
                    start += score[i][j]+score[j][i]
                elif not visited[i] and not visited[j]:
                    rink += score[i][j]+score[j][i]
        # 점수차 역대 최소인지
        answer = min(answer, abs(start-rink))
        return

    # 모든 자식 노드에 대해
    # idx = 마지막에 팀에 넣은 idx
    # 여기서 idx부터 안해주면 불필요한 것을 탐색하므로 시간초과남.
    for i in range(idx, N):
        # 유망하다면 (스타트 팀에 없다면)
        if not visited[i]:
            visited[i]=True
            # 백트래킹(n+1)
            backtracking(n+1, i+1)
            # 부모노드로 이동
            visited[i]=False

backtracking(0,0)
print(answer)

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.cc
#include <bits/stdc++.h>
using namespace std;

int N, ans = 1e9;
int S[20][20];
bool visited[20];

void backtracking(int idx, int depth){
    if (depth==N/2){
        int start = 0, rink = 0;
        for (int i=0; i<N; i++){
            for (int j=i+1; j<N; j++){
                if (visited[i] && visited[j]){
                    start += S[i][j] + S[j][i];
                }
                else if (!visited[i] && !visited[j]){
                    rink += S[i][j] + S[j][i];
                }
            }
        }
        if (abs(start-rink) < ans) {
            ans = abs(start-rink);
        }
    }

    // 마지막으로 추가한 idx+1 부터 탐색
    for (int i=idx+1; i<N; i++){
        if (visited[i]) continue;
        visited[i] = true;
        backtracking(i, depth+1);
        visited[i] = false;
    }
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(0);

    cin >> N;
    for (int i=0; i<N; i++){
        for (int j=0; j<N; j++){
            cin >> S[i][j];
        }
    }
    backtracking(-1, 0);
    cout << ans;

    return 0;
}